### PR TITLE
Catch exception thrown by HttpConnection constructor and close connection

### LINKED ISSUE #32280: HttpListener crashes application on Linux when SSL connection fails
When I create HttpListener for https (using reflection since it is not supported yet - #14691) and connect to it via http (or client just does not trust the certificate), the exception is thrown from thread outside my control and crashes whole application.

This bug is known from mono https://bugzilla.xamarin.com/show_bug.cgi?id=52675 and was fixed there https://github.com/mono/mono/pull/5724/commits/c1ddee126374599b796b6d5f3f948ddeff77f085. But it reappeared here.

Attached is [test project](https://github.com/dotnet/corefx/files/2380231/bug-test.zip), just run with "dotnet run SslConnectionCrashBugDemo.csproj". If you connect via SSL and trust the certificate (curl -k -D - -s https://localhost:4040/), it works fine - returns response and exists. But if you connect via plain HTTP, the application crashes (curl -D - -s http://localhost:4040/):

```
Unhandled Exception: System.IO.IOException: The handshake failed due to an unexpected packet format.
   at System.Net.Security.SslState.StartReadFrame(Byte[] buffer, Int32 readBytes, AsyncProtocolRequest asyncRequest)
   at System.Net.Security.SslState.StartReceiveBlob(Byte[] buffer, AsyncProtocolRequest asyncRequest)
   at System.Net.Security.SslState.ForceAuthentication(Boolean receiveFirst, Byte[] buffer, AsyncProtocolRequest asyncRequest)
   at System.Net.Security.SslState.ProcessAuthentication(LazyAsyncResult lazyResult)
   at System.Net.Security.SslStream.AuthenticateAsServer(SslServerAuthenticationOptions sslServerAuthenticationOptions)
   at System.Net.Security.SslStream.AuthenticateAsServer(X509Certificate serverCertificate, Boolean clientCertificateRequired, SslProtocols enabledSslProtocols, Boolean checkCertificateRevocation)
   at System.Net.HttpConnection..ctor(Socket sock, HttpEndPointListener epl, Boolean secure, X509Certificate cert)
   at System.Net.HttpEndPointListener.ProcessAccept(SocketAsyncEventArgs args)
   at System.Net.HttpEndPointListener.OnAccept(Object sender, SocketAsyncEventArgs e)
   at System.Net.Sockets.SocketAsyncEventArgs.OnCompleted(SocketAsyncEventArgs e)
   at System.Net.Sockets.SocketAsyncEventArgs.ExecutionCallback(Object state)
   at System.Threading.ExecutionContext.RunInternal(ExecutionContext executionContext, ContextCallback callback, Object state)
--- End of stack trace from previous location where exception was thrown ---
   at System.Net.Sockets.SocketAsyncEventArgs.FinishOperationAsyncSuccess(Int32 bytesTransferred, SocketFlags flags)
   at System.Net.Sockets.SocketAsyncEventArgs.CompletionCallback(Int32 bytesTransferred, SocketFlags flags, SocketError socketError)
   at System.Net.Sockets.SocketAsyncEventArgs.AcceptCompletionCallback(IntPtr acceptedFileDescriptor, Byte[] socketAddress, Int32 socketAddressSize, SocketError socketError)
   at System.Net.Sockets.SocketAsyncContext.AcceptOperation.InvokeCallback(Boolean allowPooling)
   at System.Net.Sockets.SocketAsyncContext.OperationQueue`1.ProcessAsyncOperation(TOperation op)
   at System.Net.Sockets.SocketAsyncContext.OperationQueue`1.<>c.<.cctor>b__18_0(Object op)
   at System.Threading.ThreadPoolWorkQueue.Dispatch()
```

### PR DESCRIPTION
**Pull request by martin-frydl**
It's a port from Mono where the issue is fixed (https://github.com/mono/mono/blob/master/mcs/class/System/System.Net/EndPointListener.cs#L117).

Fixes #32280

### CODE REVIEW

File: src/System.Net.HttpListener/src/System/Net/Managed/HttpEndPointListener.cs [hunk 1]
```diff
@@ -116,7 +116,16 @@ private static void ProcessAccept(SocketAsyncEventArgs args)
...
+            HttpConnection conn;
+            try
+            {
+                conn = new HttpConnection(accepted, epl, epl._secure, epl._cert);
+            }
+            catch
+            {
+                accepted.Close();
+                return;
+            }
```
Comment (stephentoub) [inline src/System.Net.HttpListener/src/System/Net/Managed/HttpEndPointListener.cs]: Nit: please add a blank line below the brace

**Comment by stephentoub** [comment]
Thank you for porting the fix.

Two things:
1)  Can you please add a corresponding test that would have crashed before the fix and now doesn't?
2)  The managed code base for HttpListener was originally ported from Mono ~2 years ago.  Have other fixes gone into that code that we should ensure are also ported over?

cc: @karelz

**Comment by martin-frydl** [comment]
>     1. Can you please add a corresponding test that would have crashed before the fix and now doesn't?

I can try but I'm not sure whether it will work correctly in tests. The original (old) behavior just crashed whole running process so it will probably crash and skip all following tests. I'm not sure how exactly the tests are run.

>     2. The managed code base for HttpListener was originally ported from Mono ~2 years ago.  Have other fixes gone into that code that we should ensure are also ported over?

I don't know. I'm not Mono developer, I just run our application on Mono before porting it to .NET Core now. I've checked the bugs we found in Mono implementation before and they all seem to be fixed in .NET Core version, except this one.

**Comment by stephentoub** [comment]
> I can try but I'm not sure whether it will work correctly in tests. The original (old) behavior just crashed whole running process so it will probably crash and skip all following tests. I'm not sure how exactly the tests are run.

That's fine... with your fix it shouldn't crash, and if it does crash, that'll highlight something else that needs to be fixed.

> I just run our application on Mono before porting it to .NET Core now. I've checked the bugs we found in Mono implementation before and they all seem to be fixed in .NET Core version, except this one.

Ok.  Thank you for checking.

**Comment by karelz** [comment]
@stephentoub do you think it is a good idea to accept a test which uses Reflection on our code base?
I fear it might encourage more developers to use that pattern, which is not something we want IMO.
I would rather rely on fact that nothing is broken (no regression) + manual local test. Thoughts?

**Comment by stephentoub** [comment]
> do you think it is a good idea to accept a test which uses Reflection on our code base?

In general, no.  Why would we do that?  I don't see why that's needed here.

**Comment by karelz** [comment]
@stephentoub the code that is changed here cannot be reached without tricking BCL via Reflection at this moment - see https://github.com/dotnet/corefx/issues/32280#issuecomment-421440910

**Comment by stephentoub** [comment]
> the code that is changed here cannot be reached without tricking BCL via Reflection at this moment

I see.  Then I think it's fine to just change it without adding a test.  The code is just ported from Mono, so if it's currently effectively dead code in corefx, the top priority should be keeping it in sync with the Mono code, which as I understand it has this try/catch.

**Comment by karelz** [comment]
@martin-frydl if you can confirm that this change helps your scenario (with using Reflection) locally, we are ready to merge.
Thanks for porting it over!

**Comment by martin-frydl** [comment]
Yes, I can confirm this change fixes the problems.
Thank you for merging.

**Comment by karelz** [comment]
Thanks for confirmation and for the PR @martin-frydl!

**Comment by pavelchuchma** [comment]
Hi @karelz, please, is there a way to move this fix to 2.2 release? Martin did this fix to solve existing issue in our product going to be released in next weeks. Waiting several months for Core 3.0 release of doesn't help us too much :-(

**Comment by karelz** [comment]
@pavelchuchma given that it is a change in unsupported code path and the demand is very low, it will not make it into servicing release of 2.1/2.2.

Your best options in the meantime are: Use 3.0 daily builds (without quality/support guarantees), or patch it yourself and use private builds off 2.2/2.1, or take the code and compile it into your app/library.

**Pull request merged**